rpc/oauth/web_oauth.go

```diff
@@ -0,0 +1,144 @@
+	"github.com/golang-jwt/jwt/v4"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"go.viam.com/utils/jwks"
+	"go.viam.com/utils/rpc"
+)
+
+const (
+	// CredentialsTypeOAuthWeb for jwt access tokens signed by oidc/oauth backend.
+	CredentialsTypeOAuthWeb = rpc.CredentialsType("oauth-web-auth")
+)
+
+type webOAuthHandler struct {
+	WebOAuthOptions
+}
+
+var (
+	_ rpc.AuthHandler              = &webOAuthHandler{}
+	_ rpc.TokenCustomClaimProvider = &webOAuthHandler{}
+)
+
+type webOAuthClaims struct {
+	rpc.JWTClaims
+
+	// The JWT must contain the aud claim otherwise it must be rejected.
+	allowedAudience string
+}
+
+func (c *webOAuthClaims) Entity() (string, error) {
+	// Claim `sub` will contain the Auth0 userid not the basic email.
+	if email, ok := c.AuthMetadata["email"]; ok {
+		return email, nil
+	}
+
+	return "", errors.New("missing email in rpc_auth_md")
+}
+
+func (c *webOAuthClaims) Valid() error {
+	if !c.RegisteredClaims.VerifyAudience(c.allowedAudience, true) {
+		return errors.New("invalid aud")
+	}
+
+	return c.JWTClaims.Valid()
+}
+
+var _ rpc.Claims = &webOAuthClaims{}
+
+// WebOAuthOptions options for the WebOauth handler.
+type WebOAuthOptions struct {
+	// Audience claim that must be within the "aud" JWT claims presented.
+	AllowedAudience string
+
+	// Key provider used to provide public keys to validate the jwt basid on its "kid" header.
+	KeyProvider jwks.KeyProvider
+
+	// Underlying Entity verifier after
+	EntityVerifier func(ctx context.Context, entity string) (interface{}, error)
+	Logger         golog.Logger
+}
+
+// WithWebOAuthTokenAuthHandler returns a rpc server option configured for the AuthHandler. The WebAuth handler will
+// validate jwt access tokens signed by OIDC provider. The jwts are validated for the aud standard claim.
+//
+// This allows auth handler allows for a dynamic set of signing keys provided through OIDC configuration endpoint.
+//
+// Entity verification is deleged to the EntityVerifier method in the options.
+func WithWebOAuthTokenAuthHandler(opts WebOAuthOptions) rpc.ServerOption {
+	authHandler := &webOAuthHandler{
+		opts,
+	}
+
+	return rpc.WithAuthHandler(CredentialsTypeOAuthWeb, authHandler)
+}
+
+// Authenticate always returns an error. Webauth is expecting an access token to be generated by a separate
+// system instead of using Authenticate() to create the access token.
+func (a *webOAuthHandler) Authenticate(ctx context.Context, entity, payload string) (map[string]string, error) {
+	return nil, status.Error(codes.Unimplemented, "not implemented with webauth")
+}
+
+// VerifyEntity verifies that this handler is allowed to authenticate the given entity.
+// The handler can optionally return opaque info about the entity that will be bound to the
+// context accessible via ContextAuthEntity.
+func (a *webOAuthHandler) VerifyEntity(ctx context.Context, entity string) (interface{}, error) {
+	return a.EntityVerifier(ctx, entity)
```

Comment:
if this is nil, should we error?

Reply:
Yeah good call. Throwing an internal error and added tests.